File: 01-day/main.py
with open('input.txt', 'r') as f:
    lines = f.readlines()

measurements = [int(line.strip()) for line in lines]

increases = [
    1
    for a, b in zip(measurements[:-1], measurements[1:])
    if b - a > 0
]

n_increases = sum(increases)
print(f"A ::: Number of increases: {n_increases}")

# ============================================================================

lag0 = measurements[:-2]
lag1 = measurements[1:-1]
lag2 = measurements[2:]

measurements_lagged = [a + b + c for a, b, c in zip(lag0, lag1, lag2)]

increases_lagged = [
    1
    for a, b in zip(measurements_lagged[:-1], measurements_lagged[1:])
    if b - a > 0
]

n_increases_lagged = sum(increases_lagged)
print(f"B ::: Number of increases: {n_increases_lagged}")

File: 01-day/main.scala
import scala.io.Source

val data = Source.fromFile("input.txt").getLines.toArray.map(_.toInt)

val n_increases = (data.dropRight(1) zip data.drop(1))
  .filter(t => t._2 - t._1 > 0)
  .length

println(n_increases)

val lag0 = data.dropRight(2)
val lag1 = data.drop(1).dropRight(1)
val lag2 = data.drop(2)

val data_rolling = (lag0, lag1, lag2).zipped.toList.map(t => t._1 + t._2 + t._3)
val n_increases_rolling = (data_rolling.dropRight(1) zip data_rolling.drop(1))
  .filter(t => t._2 - t._1 > 0)
  .length

println(n_increases_rolling)
